File: app/src/main/java/com/jiujiu/mvvmTemplate/di/MyWorkerFactory.java
package com.jiujiu.mvvmTemplate.di;

import android.content.Context;

import com.jiujiu.mvvmTemplate.worker.CustomWorkerFactory;

import java.util.Map;

import javax.inject.Inject;
import javax.inject.Provider;
import javax.inject.Singleton;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.work.ListenableWorker;
import androidx.work.WorkerFactory;
import androidx.work.WorkerParameters;

@Singleton
public class MyWorkerFactory extends WorkerFactory {

    private final Map<Class<? extends ListenableWorker>, Provider<CustomWorkerFactory>> workerFactories;

    @Inject
    public MyWorkerFactory(Map<Class<? extends ListenableWorker>, Provider<CustomWorkerFactory>> workerFactories) {
        this.workerFactories = workerFactories;
    }

    @Nullable
    @Override
    public ListenableWorker createWorker(@NonNull Context appContext, @NonNull String workerClassName, @NonNull WorkerParameters workerParameters) {
        ListenableWorker listenableWorker = null;
        try {
            Provider<? extends CustomWorkerFactory> provider = null;
            for (Map.Entry<Class<? extends ListenableWorker>, Provider<CustomWorkerFactory>> entry : workerFactories.entrySet()) {
                if (Class.forName(workerClassName).isAssignableFrom(entry.getKey())) {
                    provider = entry.getValue();
                    break;
                }
            }
            if (provider == null) {
                throw new IllegalArgumentException("unknown model class " + workerClassName);
            }
            listenableWorker = provider.get().create(appContext, workerParameters);
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
        return listenableWorker;
    }
}

File: app/src/main/java/com/jiujiu/mvvmTemplate/di/MyWorkerFactory.kt
package com.jiujiu.mvvmTemplate.di

import android.content.Context
import androidx.work.ListenableWorker
import androidx.work.WorkerFactory
import androidx.work.WorkerParameters

import com.jiujiu.mvvmTemplate.worker.CustomWorkerFactory

import javax.inject.Inject
import javax.inject.Provider
import javax.inject.Singleton

@Singleton
class MyWorkerFactory @Inject
constructor(private val workerFactories: MutableMap<Class<out ListenableWorker>, Provider<CustomWorkerFactory>>) : WorkerFactory() {

    override fun createWorker(appContext: Context, workerClassName: String, workerParameters: WorkerParameters): ListenableWorker? {
        var listenableWorker: ListenableWorker? = null
        try {
            var provider: Provider<out CustomWorkerFactory>? = null
            for ((key, value) in workerFactories) {
                if (Class.forName(workerClassName).isAssignableFrom(key)) {
                    provider = value
                    break
                }
            }
            if (provider == null) {
                throw IllegalArgumentException("unknown model class $workerClassName")
            }
            listenableWorker = provider.get().create(appContext, workerParameters)
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }

        return listenableWorker
    }
}
